### PR TITLE
refactor(engine-core): allow list `ElementInternals` APIs and update type definitions

### LINKED ISSUE #3782: Follow-up work for ElementInternals
I think there are a few things we missed in #3670:

1. The types should be updated [here](https://github.com/salesforce/lwc/blob/54c64e29d420b2356ab4e58763cbf5bae0b83f07/packages/lwc/index.d.ts) to add `this.attachInternals` for those using TypeScript.
2. For the Proxy on the `ElementInternals` object, we should maintain an allowlist of known methods/props that the user is allowed to access. Otherwise browsers may implement new methods/properties in the future which would break our simulation of `formAssociated` and allow access to form-related things that the component should not have access to. (Given that ARIA reflection is implemented unevenly across browsers, we can use a regex like `/^aria/` for those props.)

### PR DESCRIPTION
**Pull request by jmsjtu**
## Details
Fixes #3782 

Updated `ElementInternals` proxy to an allow listed set of APIs.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## GUS work item
W-14260034

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts [hunk 1]
```diff
@@ -335,19 +364,31 @@ function createElementInternalsProxy(
...
-                ? internalsPropertyValue.bind(target)
-                : internalsPropertyValue;
+            if (
+                // Pass through Object.prototype methods such as toString()
+                hasOwnProperty.call(Object.prototype, propertyKey) ||
+                // As of this writing, ElementInternals only uses Symbol.toStringTag which is called
+                // on Object.hasOwnProperty invocations
+                Symbol.for('Symbol.toStringTag') === propertyKey ||
+                // ElementInternals allow listed properties
+                isAllowedElementInternalAccessor(propertyKey)
```
Comment (jmsjtu) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:383]: I check here if the accessor is calling into a function from `Object.prototype` before validating against the allow list of properties.

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts [hunk 2]
```diff
@@ -315,15 +316,41 @@ const formAssociatedProps = new Set([
 ]);
 
 // Verify that access to a form-associated property of the ElementInternals proxy has formAssociated set in the LWC.
-function assertFormAssociatedPropertySet(propertyKey: string, isFormAssociated: boolean) {
-    if (formAssociatedProps.has(propertyKey) && !isFormAssociated) {
+function verifyPropForFormAssociation(propertyKey: string | symbol, isFormAssociated: boolean) {
+    if (
+        typeof propertyKey == 'string' &&
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:321]:
```suggestion
        isString(propertyKey)
```

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts [hunk 3]
```diff
@@ -315,15 +316,41 @@ const formAssociatedProps = new Set([
...
+const elementInternalsAccessorAllowList = new Set(['shadowRoot', 'role', ...formAssociatedProps]);
+
+// Prevent access to properties not defined in the HTML spec in case browsers decide to
+// provide new APIs that provide access to form associated properties.
+// This can be removed along with UpgradeableConstructor.
+function isAllowedElementInternalAccessor(propertyKey: string | symbol) {
+    let isAllowedAccessor = false;
+    // As of this writing all ElementInternal property keys as described in the spec are implemented with strings
+    // in Chrome, Firefox, and Safari
+    if (typeof propertyKey === 'string') {
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:341]:
```suggestion
    if (isString(propertyKey)) {
```

File: packages/@lwc/engine-core/src/framework/base-lightning-element.ts [hunk 4]
```diff
@@ -335,19 +362,31 @@ function createElementInternalsProxy(
...
+                // As of this writing, ElementInternals only uses Symbol.toStringTag which is called
+                // on Object.hasOwnProperty invocations
+                Symbol.for('Symbol.toStringTag') === propertyKey ||
+                // ElementInternals allow listed properties
+                isAllowedElementInternalAccessor(propertyKey)
+            ) {
+                // Verify that formAssociated is set for form associated properties
+                verifyPropForFormAssociation(propertyKey, isFormAssociated);
+                const propertyValue = Reflect.get(target, propertyKey);
+                return typeof propertyValue === 'function'
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/base-lightning-element.ts:386]:
```suggestion
                return isFunction(propertyValue)
```

**Review by nolanlawson — APPROVED** [review]
LGTM!

**Pull request merged**